lib/services/notifications.go

```diff
@@ -0,0 +1,319 @@
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// Notifications defines an interface for managing notifications.
+type Notifications interface {
+	NotificationsGetter
+	CreateUserNotification(ctx context.Context, username string, notification *notificationsv1.Notification) error
+	DeleteUserNotification(ctx context.Context, username string, notificationId string) error
+	DeleteAllUserNotificationsForUser(ctx context.Context, username string) error
+	CreateGlobalNotification(ctx context.Context, globalNotification *notificationsv1.GlobalNotification) (notificationsv1.GlobalNotification, error)
+	DeleteGlobalNotification(ctx context.Context, notificationId string) error
+	UpsertUserNotificationState(ctx context.Context, username string, state *notificationsv1.UserNotificationState) (notificationsv1.UserNotificationState, error)
+	DeleteUserNotificationState(ctx context.Context, username string, notificationId string) error
+	DeleteAllUserNotificationStatesForUser(ctx context.Context, username string) error
+	ListUserNotificationStates(ctx context.Context, username string, pageSize int, nextToken string) ([]notificationsv1.UserNotificationState, string, error)
+	UpsertUserLastSeenNotification(ctx context.Context, username string, ulsn *notificationsv1.UserLastSeenNotification) (notificationsv1.UserLastSeenNotification, error)
+	GetUserLastSeenNotification(ctx context.Context, username string) (notificationsv1.UserLastSeenNotification, error)
+	DeleteUserLastSeenNotification(ctx context.Context, username string) error
+}
+
+// NotificationsGetter is an interface for fetching Notifications resources.
+type NotificationsGetter interface {
+	ListNotificationsForUser(context.Context, notificationsv1.ListUserNotificationsRequest) ([]notificationsv1.Notification, error)
+}
+
+// ValidateNotification verifies that the necessary fields are configured before a notification can be created.
+func ValidateNotification(notification *notificationsv1.Notification) error {
+	if notification.SubKind == "" {
+		return trace.BadParameter("notification subkind is missing")
+	}
+
+	if notification.Metadata == nil {
+		return trace.BadParameter("notification metadata is missing")
+	}
+
+	if notification.Metadata.Labels == nil {
+		return trace.BadParameter("notification metadata labels are missing")
+	}
+
+	return nil
+}
+
+// MarshalNotification marshals a Notification resource to JSON.
+func MarshalNotification(notification *notificationsv1.Notification, opts ...MarshalOption) ([]byte, error) {
+	err := ValidateNotification(notification)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !cfg.PreserveResourceID {
+		notification = proto.Clone(notification).(*notificationsv1.Notification)
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		notification.Metadata.Id = 0
+		notification.Metadata.Revision = ""
+	}
```

Comment:
It doesn't look like we can because `maybeResetProtoResourceID` requires the resource type to have the `SetResourceID()` and `SetRevision()` methods, which `notificationsv1.Notification` doesn't have, we would need to clone it and add them which I don't think is worth doing for this.